File: agent/library/src/main/java/org/matilda/di/destructors/DestructionManager.java
package org.matilda.di.destructors;

import javax.inject.Inject;
import javax.inject.Singleton;
import java.util.Stack;

@Singleton
public class DestructionManager {
    private final Stack<Destructor> mDestructors;

    @Inject
    public DestructionManager() {
        mDestructors = new Stack<>();
    }

    public void addDestructor(Destructor destructor) {
        mDestructors.push(destructor);
    }

    public void destruct() {
        while (!mDestructors.empty()) {
            mDestructors.pop().destruct();
        }
    }
}

File: agent/library/src/main/java/org/matilda/di/destructors/DestructionManager.kt
package org.matilda.di.destructors

import me.tatarka.inject.annotations.Inject
import org.matilda.commands.MatildaScope
import java.util.*

@MatildaScope
@Inject
class DestructionManager {
    private val mDestructors = Stack<Destructor>()

    fun addDestructor(destructor: Destructor) {
        mDestructors.push(destructor)
    }

    fun destruct() {
        while (!mDestructors.empty()) {
            mDestructors.pop().destruct()
        }
    }
}
